AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java

```diff
@@ -268,29 +287,41 @@ protected TaskData doInBackground(TaskData... params) {
         }
     }
 
+    /** Ensure all previous tasks are finished.
+     * @return whether computation can occur*/
+    private synchronized boolean allPreviousTasksDoneCorrectly() {
```

Comment:
This represents a continued misunderstanding of java synchronization. The synchronized marker on a method says "automatically grab the class lock for static methods, or the instance lock for instance methods"

So it is not using your new lock object at all, and should be removed

The code should be in a `synchronized (lock) {  /* all the code ... */ }` block to get the locking semantics you want.

Reply:
I thought I removed it. I forgot I split the method in two and that it became useless

Reply:
this whole code block does not seem to be accessing the things I thought were touched by multiple threads, so this should not be synchronized marker at all? And even so it should not have the method marker as we using an explicit lock object now

```suggestion
    private boolean allPreviousTasksDoneCorrectly() {
```

Reply:
I'll answer one last time and then if it does not make sens, I'll quit for today. I fear I'm just too tired to make anything more complex than pasta correctly right now.
My previous comment about synchronization considered the two static variable and totally forgot to consider this method.

The static variable is synchronized in order to ensure that we are dealing with a list and not a tree.

The method allPreviousTaskDoneCorrectly is entirely independent. In this case, my goal is really to ensure that one task is locked until all previous tasks are done. I need this method to be synchronized to ensure that no too thread are modifying `mPreviousTask` concurrently. Even if I don't believe it would be possible anyway, since the method is supposed to run in its own thread, and so it should never be called twice by a method